lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,407 @@
+		ret = sample_rate_converter_reconfigure(ctx, input_sample_rate, output_sample_rate,
+							filter);
+		if (ret) {
+			LOG_ERR("Failed to initialize converter (%d)", ret);
+			return ret;
+		}
+	}
+
+	if ((ctx->conversion_direction == CONVERSION_DIR_DOWN) &&
+	    (samples_in < ctx->conversion_ratio)) {
+		LOG_ERR("Number of samples in can not be less than the conversion ratio (%d) when "
+			"downsampling",
+			ctx->conversion_ratio);
+		return -EINVAL;
+	}
+
+	if (ctx->conversion_direction == CONVERSION_DIR_UP) {
+		*output_written = input_size * ctx->conversion_ratio;
+	} else {
+		*output_written = input_size / ctx->conversion_ratio;
+	}
+
+	if (*output_written > output_size) {
+		LOG_ERR("Conversion process will produce more bytes than the output buffer can "
+			"hold");
+		return -EINVAL;
+	}
+
+	if (*output_written > SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE) {
+		LOG_ERR("Conversion process will produce more bytes than the internal output "
+			"buffer can hold");
+		return -EINVAL;
+	}
+
+	if (!conversion_needs_buffering(ctx, ctx->conversion_ratio)) {
+		write_ptr = output;
+		read_ptr = input;
+		samples_to_process = samples_in;
+	} else {
+		read_ptr = internal_input_buf;
+		write_ptr = internal_output_buf;
+
+		if (((samples_in + (ctx->input_buf.bytes_in_buf * bytes_per_sample)) %
+		     ctx->conversion_ratio) == 0) {
+			size_t extra_samples =
+				ctx->conversion_ratio - (samples_in % ctx->conversion_ratio);
+
+			LOG_DBG("Using %d extra samples from input buffer", extra_samples);
+			samples_to_process = samples_in + extra_samples;
+		} else {
+			size_t extra_samples = (samples_in % ctx->conversion_ratio);
+
+			LOG_DBG("Storing %d samples in input buffer for next iteration",
+				extra_samples);
+			samples_to_process = samples_in - extra_samples;
+		}
+
+		/* Merge bytes in input buffer and incoming bytes into the internal buffer for
+		 * processing
+		 */
+		memcpy(internal_input_buf, ctx->input_buf.buf, ctx->input_buf.bytes_in_buf);
+		memcpy(internal_input_buf + ctx->input_buf.bytes_in_buf, input, input_size);
+	}
+
+#ifdef CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_16
+	if (ctx->conversion_direction == CONVERSION_DIR_UP) {
+		arm_fir_interpolate_q15(&ctx->fir_interpolate_q15, (q15_t *)read_ptr,
+					(q15_t *)write_ptr, samples_to_process);
+	} else {
+		arm_fir_decimate_q15(&ctx->fir_decimate_q15, (q15_t *)read_ptr, (q15_t *)write_ptr,
+				     samples_to_process);
+	}
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+	if (ctx->conversion_direction == CONVERSION_DIR_UP) {
+		arm_fir_interpolate_q31(&ctx->fir_interpolate_q31, (q31_t *)read_ptr,
+					(q31_t *)write_ptr, samples_to_process);
+	} else {
+		arm_fir_decimate_q31(&ctx->fir_decimate_q31, (q31_t *)read_ptr, (q31_t *)write_ptr,
+				     samples_to_process);
+	}
+#endif
+
+	if (!conversion_needs_buffering(ctx, ctx->conversion_ratio)) {
+		/* Nothing needs to be done in output buffer */
+		return 0;
+	}
+
+	if (samples_to_process < samples_in) {
```

Comment:
Would appreciate a walkthrough of the last part of this function.

Reply:
Sure, let's do that offline. I was also surprised by how long this last part ended up being